### ISSUE TITLE
More gentle intro intro grammers.

### ISSUE THREAD

**Issue by berry-thawson** [munificent/craftinginterpreters#553]
I am guy from non-cs background, and recently stumbled upon your book. So, this might be different for you folks.
Before starting on yours, I came across various other resources like - Dick Grune's "Parsing techniques" which was super fun, but seems to be unconsciously glossing over things.
Like when explaining about grammar and different types of grammar - he just explains context 
sensitive grammars as :

> A grammar is Type 1 monotonic if it contains no rules in which the left-hand side
> consists of more symbols than the right-hand side. This forbids, for instance, the rule ,
> N E - > a n d N .
> A grammar is Type 1 context-sensitive if all of its rules are context-sensitive. A
> rule is context-sensitive if actually only one (non-terminal) symbol in its left-hand side
> gets replaced by other symbols, while we find the others back undamaged and in the
> same order in the right-hand side. Example:
> Name Comma Name End - > Name and Name End
> which tells that the rule
> Comma - > and
> may be applied if the left context is N a m e and the right context is N a m e E n d . The con-
> texts themselves are not affected. The replacement must be at least one symbol long;
> this means that context-sensitive grammars are always monotonic; see Section 2.6.

This was too abrupt because what he means here as context was not explained here at all. Only when we come to context free part, we get some idea about this:

> Type 2 grammars are called context-free grammars (CF grammars) and their relation to
> context-sensitive grammars is as direct as the name suggests. A context-free grammar
> is like a context-sensitive grammar, except that both the left and the right contexts are
> required to be absent (empty). As a result, the grammar may contain only rules that
> have a single non-terminal on their left-hand side.
OK. Context free grammars are one when LHS has like only one non-terminal, so we are dependent on look ahead or look behind operations before performing replacements.

Though journey was tedious, it was fun. 

Your intro about AST was too good, but it was odd that terms like formal grammar, Chomsky hierarchy...  see just sprinkled over beginning of chapter. And reference links were wikipedia articles?

Though it would  lengthen the chapter, please don't give wikipedia pages as intro to grammars. That's can leave noobs like me bewildered.

Thanks for your book. It helps me a lot. Being an guy who never read CS, it was great intro to things from your field. Can you recommend some other books too?

**Comment by munificent** [comment]
This is a fair request. One of the biggest challenges in writing a book is deciding what *not* to include. A better explanation of the Chomsky hierarchy and formal languages would be useful for some, but would also add another 20,000 words of prose that may not be that interesting to many readers. As I say in the previous chapter:

> It pains me to gloss over the theory so much, especially when it’s as fun as I think the Chomsky hierarchy and FSMs are. But the honest truth is other books cover this better than I could. Compilers: Principles, Techniques, and Tools (universally known as “the Dragon Book”) is the canonical reference.

In general, my book is theory-light. My main hope is to give you a concrete intuition about languages which will prepare you to absorb deeper theoretical books later. If I pack too much theory into this book, fewer readers will make it through.

**Comment by berry-thawson** [comment]
I feel kind of stupid for mentioning "Parsing techniques". I was trying to explain that, this book was very vague in expressing what is context and how context sensitive which I wish did not face as a total beginner without any guidance. I went ahead after read about parsing after reading through your book. Thanks for that.

Yeah, I get what you are trying to do.

I had to pause and refer other books precisely at context-free part, because it has been explained very nicely as a useful tool, but term context-free itself was vague. 

Maybe, I approached things with extra caution due to lack of my background.

My only suggestion is if some term like "context-free" is mentioned, it should be explained.What is context? How does it feel to be finally free of "context"? It is just that you don't have to do look ahead and look behind before doing the replacement for generating sentences from grammar, (I guess).

Thanks for your reply. I am doing xpath like query language with your book as reference.

You are really doing great job with the book. 

Cheers.
